networkx/algorithms/isomorphism/vf2pp.py

```diff
@@ -0,0 +1,306 @@
+    _restore_Tinout,
+    _update_Tinout,
+)
+
+__all__ = ["vf2pp_mapping", "vf2pp_is_isomorphic", "vf2pp_all_mappings"]
+
+_GraphParameters = collections.namedtuple(
+    "_GraphParameters",
+    [
+        "G1",
+        "G2",
+        "G1_labels",
+        "G2_labels",
+        "nodes_of_G1Labels",
+        "nodes_of_G2Labels",
+        "G2_nodes_of_degree",
+    ],
+)
+_StateParameters = collections.namedtuple(
+    "_StateParameters",
+    ["mapping", "reverse_mapping", "T1", "T1_out", "T2", "T2_out"],
+)
+
+
+def vf2pp_mapping(G1, G2, node_labels=None, default_label=None):
+    """Return an isomorphic mapping between `G1` and `G2` if it exists.
+
+    Parameters
+    ----------
+    G1,G2: NetworkX Graph or MultiGraph instances.
+        The two graphs to check for isomorphism or monomorphism.
+
+    node_labels: Label name
+        The label name of all nodes
+
+    default_label: Label name
+        Let the user pick a default label value
+
+    Returns
+    -------
+    Node mapping, if the two graphs are isomorphic. None otherwise.
+    """
+    try:
+        mapping = next(vf2pp_all_mappings(G1, G2, node_labels, default_label))
+        return mapping
+    except StopIteration:
+        return None
+
+
+def vf2pp_is_isomorphic(G1, G2, node_labels=None, default_label=None):
+    """Examines whether G1 and G2 are isomorphic.
+
+    Parameters
+    ----------
+    G1,G2: NetworkX Graph or MultiGraph instances.
+        The two graphs to check for isomorphism or monomorphism.
+
+    node_labels: Label name
+        The label name of all nodes
+
+    default_label: Label name
+        Let the user pick a default label value
+
+    Returns
+    -------
+    True if the two graphs are isomorphic. False otherwise.
+    """
+    if vf2pp_mapping(G1, G2, node_labels, default_label) is not None:
+        return True
+    return False
+
+
+def vf2pp_all_mappings(G1, G2, node_labels=None, default_label=None):
+    """Yields all the possible mappings between G1 and G2.
+
+    Parameters
+    ----------
+    G1,G2: NetworkX Graph or MultiGraph instances.
+        The two graphs to check for isomorphism.
+
+    node_labels: string or None
+        The node attribute name within G that indicates node labels.
+        If None, then no node labels are used to compute isomorphisms.
+
+    default_label: string
+        The default label for nodes that have no label attribute value.
+    """
+    if G1.number_of_nodes() == 0 or G2.number_of_nodes() == 0:
+        return False
+
+    # Check that both graphs have the same number of nodes and degree sequence
+    if not _fast_precheck(G1, G2):
+        return False
+
+    # Initialize parameters and cache necessary information about degree and labels
+    G1_labels = dict(G1.nodes(data=node_labels, default=default_label))
+    G2_labels = dict(G2.nodes(data=node_labels, default=default_label))
+    graph_params, state_params = _initialize_parameters(G1, G2, G1_labels, G2_labels)
```

Comment:
I think you should put the creation of `G1_labels` and `G2_labels` into the `_initialize_parameters` function. That isolates the data structure creation into one function. The possible advantage of that is to simplify constructing gparams and sparams in the tests. 

Would it help simplify the tests to make the initialize function take inputs `_initialize_parameters(G1, G2, node_labels, default_label)`? Then you could remove the definition of `_GraphParameters` from the test functions as well as construction of T1 and T2, etc. Tests like `test_updating` could start with something like `gparams, sparams = _initialize_parameters(self.G1, self.G2, None, None)` or similar.

I haven't looked closely at this. But it seems like that could simplify the test setup code. If not, that's fine. Leave the code as is.  But it seems like it is worth checking. That's one of the nice things about localizing the initialization code.  But if you end up having to put even more code to set up G1 and G2 then it is not worth it.